### PR TITLE
Add android support

### PR DESCRIPTION
**Pull request by markusbug**
Many commits about adding support for android on Daimo.

Things I changed in the auto-generated android code:

- Add line at the bottom of the app-level build.gradle to apply from fonts.gradle in icons
- Add google-services to the dependencies in the project-level build.gradle
- Add google-services to the app-level build.gradle
- In the AndroidManifest added intent-filter for app deep-linking

### CODE REVIEW

File: apps/daimo-mobile/app.json [hunk 1]
```diff
@@ -30,11 +30,36 @@
...
+            {
+              "scheme": "exp+daimo"
+            },
+            {
+              "scheme": "https",
+              "host": "daimo.onrender.com",
+              "pathPrefix": "/link"
+            }
+          ]
+        }
```
Comment (dcposch) [inline apps/daimo-mobile/app.json:55]: 👍 👍

File: apps/daimo-mobile/app.json [hunk 2]
```diff
@@ -30,11 +30,36 @@
       }
     },
     "android": {
+      "googleServicesFile": "./google-services.json",
+      "useNextNotificationsApi": true,
```
[the remark this answers is not in the document]
Comment (dcposch) [reply apps/daimo-mobile/app.json:34]: wonder if it still does anything in latest expo version. recommend removing if notifications work without it

File: apps/daimo-mobile/metro.config.js [hunk 3]
```diff
@@ -1,8 +1,14 @@
...
 
 const config = getDefaultConfig(__dirname); // eslint-disable-line no-undef
 
-config.resolver = { ...config.resolver, unstable_enablePackageExports: true };
+config.resolver = {
+  ...config.resolver,
+  extraNodeModules: {
+    // Add any additional packages/modules you want to include
+  },
+  assetExts: [...config.resolver.assetExts, "ttf", "woff", "woff2"], // Include font extensions
```
Comment (dcposch) [inline apps/daimo-mobile/metro.config.js:10]: for octicons 👍

File: apps/daimo-mobile/package.json [hunk 4]
```diff
@@ -6,9 +6,12 @@
   "scripts": {
     "build:sim": "eas build -p ios --profile=sim --non-interactive",
     "build:dev": "eas build -p ios --profile=dev --non-interactive",
+    "build:dev-android": "eas build -p android --profile=dev --non-interactive",
     "build:stage": "eas build -p ios --profile=stage --auto-submit --non-interactive",
+    "build:stage-android": "eas build -p android --profile=stage --non-interactive",
     "submit:stage": "eas submit --platform ios --profile=stage",
     "dev": "expo start --dev-client --clear -i",
+    "dev:android": "expo start --dev-client --clear -a",
```
Comment (dcposch) [inline apps/daimo-mobile/package.json:14]: 👍

File: apps/daimo-mobile/package.json [hunk 5]
```diff
@@ -18,7 +21,9 @@
     "lint:style": "eslint . --max-warnings=0",
     "fix:deps": "expo install --fix",
     "fix:style": "eslint --fix .",
-    "eas-build-post-install": "cd ../.. && npm run build"
+    "eas-build-post-install": "cd ../.. && npm run build",
+    "android": "expo run:android",
+    "ios": "expo run:ios"
```
Comment (dcposch) [inline apps/daimo-mobile/package.json:26]: can delete `run:...` targets, given that expo likes to auto-add `npm run android` / `npm run ios`

File: apps/daimo-mobile/package.json [hunk 6]
```diff
@@ -49,12 +55,15 @@
...
     "expo-device": "^5.2.1",
+    "expo-font": "^11.4.0",
     "expo-linking": "~5.0.2",
+    "expo-local-authentication": "~13.4.1",
     "expo-localization": "^14.3.0",
     "expo-notifications": "^0.20.1",
     "expo-splash-screen": "~0.20.4",
     "expo-status-bar": "~1.6.0",
     "fast-text-encoding": "^1.0.6",
+    "firebase": "^10.1.0",
```
Comment (dcposch) [inline apps/daimo-mobile/package.json:66]: new dependencies:
`firebase`, `@react-native-firebase`, `expo-font`

File: apps/daimo-mobile/package.json [hunk 7]
```diff
@@ -93,5 +101,8 @@
         }
       ]
     }
+  },
+  "devDependencies": {
+    "@babel/core": "^7.20.0"
```
Comment (dcposch) [inline apps/daimo-mobile/package.json:106]: delete, already a `dependency`

File: apps/daimo-mobile/src/App.tsx [hunk 8]
```diff
@@ -20,6 +21,11 @@ export default function App() {
   // Track names for known addresses
   useEAccountCache();
 
+  // Load font to fix icons on android
+  useFonts({
+    'Octicons': require('../assets/Octicons.ttf'),
```
Comment (dcposch) [inline apps/daimo-mobile/src/App.tsx:26]: 👍

File: apps/daimo-mobile/src/logic/notify.ts [hunk 9]
```diff
@@ -62,6 +65,30 @@ class PushNotificationManager {
 
   maybeSavePushTokenForAccount = async () => {
     try {
+      if (Platform.OS === "android") {
+        // Android specific setup
+        await Notifications.setNotificationChannelAsync('default', {
+          name: 'default',
+          importance: Notifications.AndroidImportance.MAX,
+          vibrationPattern: [0, 250, 250, 250],
+          lightColor: '#FF231F7C',
```
Comment (dcposch) [inline apps/daimo-mobile/src/logic/notify.ts:74]: do `lightColor`/`vibrationPattern` have reasonable defaults?

File: apps/daimo-mobile/src/logic/notify.ts [hunk 10]
```diff
@@ -62,6 +65,30 @@ class PushNotificationManager {
...
+        });
+
+        const firebaseConfig = {
+          apiKey: googleServicesConfig.client[0].api_key[0].current_key,
+          authDomain: "daimo-98c63.firebaseapp.com",
```
Comment (dcposch) [inline apps/daimo-mobile/src/logic/notify.ts:79]: pull from `googleServicesConfig` vs hardcoding if possible

File: apps/daimo-mobile/src/logic/notify.ts [hunk 11]
```diff
@@ -62,6 +65,30 @@ class PushNotificationManager {
...
+          projectId: googleServicesConfig.project_info.project_id,
+          storageBucket: googleServicesConfig.project_info.storage_bucket,
+          messagingSenderId: "164379879968",
```
Comment (dcposch) [inline apps/daimo-mobile/src/logic/notify.ts:82]: googleServicesConfig projectNumber

File: apps/daimo-mobile/src/view/screen/receive/RequestSendScreen.tsx [hunk 12]
```diff
@@ -36,7 +36,7 @@ export default function RequestSendScreen() {
 
   const sendRequest = useCallback(async () => {
     try {
-      const result = await Share.share({ url });
+      const result = await Share.share({ message: url });
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/receive/RequestSendScreen.tsx:39]: 👍 thanks for testing in iOS sim!

File: apps/daimo-mobile/src/view/shared/Input.tsx [hunk 13]
```diff
@@ -43,6 +43,8 @@ export function InputBig({
         numberOfLines={1}
         autoCapitalize="none"
         autoCorrect={false}
+        secureTextEntry={true}
+        keyboardType={"visible-password"}
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/shared/Input.tsx:47]: 👍 nice workaround

File: apps/daimo-mobile/src/view/shared/Input.tsx [hunk 14]
```diff
@@ -70,23 +72,22 @@ export function AmountInput({
...
+    if (!(newVal >= 0)) {
+      newVal = 0;
+    }
+    const newStrVal = fmt(newVal);
+    setStrVal(newVal > 0 ? newStrVal : "");
+
+    const truncated = parseLocalFloat(newStrVal);
+    onChange(truncated);
+    if (onSubmitEditing) onSubmitEditing(truncated);
+  };
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/shared/Input.tsx:88]: 👍 👍

**Review by dcposch — COMMENTED** [review]
Looks excellent. Left comments

Big q is whether we need to generate+edit `/android/`, or if we can do those edits in Expo config upstream of generation

File: apps/daimo-mobile/app.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (markusbug) [reply apps/daimo-mobile/app.json:34]: @dcposch Ok just tested and it still works without. Removing

File: apps/daimo-mobile/src/logic/notify.ts (hunk 9 above)
Comment (markusbug) [reply apps/daimo-mobile/src/logic/notify.ts:74]: @dcposch Yeah, it's comfortable I would say

**Comment by dcposch** [comment]
## Dev quickstart

- JS development README instructions. Install Android Studio, run emulator. Download latest `Android internal distribution build` from Expo and install in emulator. Finally, `cd apps/daimo-mobile && npm run dev:android`

- Native development
  - Need the right Java version. Java 20 failed with a gradle error, Java 17 works.
  - Need to set `ANDROID_HOME`. Appears installing Android Studio doesn't do this automatically.

**Comment by dcposch** [comment]
By the way, have you tried it in ethOS?

**Comment by dcposch** [comment]
LGTM after https://github.com/markusbug/daimo/pull/2 + rebase

**Pull request merged**